coderd/autobuild/lifecycle_executor.go

```diff
@@ -218,53 +200,84 @@ func (e *Executor) runOnce(t time.Time) Stats {
+	// autostart the workspace.
+	if !templateSchedule.UserAutostartEnabled || !ws.AutostartSchedule.Valid || ws.AutostartSchedule.String == "" {
+		return false
 	}
-	// Don't check the template schedule to see whether it allows autostop, this
-	// is done during the build when determining the deadline.
-	if priorHistory.Transition == database.WorkspaceTransitionStart && !priorHistory.Deadline.IsZero() {
-		return true
+
+	sched, err := schedule.Weekly(ws.AutostartSchedule.String)
+	if err != nil {
+		return false
 	}
+	// Round down to the nearest minute, as this is the finest granularity cron supports.
+	// Truncate is probably not necessary here, but doing it anyway to be sure.
+	nextTransition := sched.Next(build.CreatedAt).Truncate(time.Minute)
 
-	return false
+	return !currentTick.Before(nextTransition)
 }
 
-func getNextTransition(
-	ws database.Workspace,
-	priorHistory database.WorkspaceBuild,
-	priorJob database.ProvisionerJob,
-) (
-	validTransition database.WorkspaceTransition,
-	nextTransition time.Time,
-	err error,
-) {
-	if !priorJob.CompletedAt.Valid || priorJob.Error.String != "" {
-		return "", time.Time{}, xerrors.Errorf("last workspace build did not complete successfully")
+// isEligibleForAutostart returns true if the workspace should be autostopped.
+func isEligibleForAutostop(build database.WorkspaceBuild, job database.ProvisionerJob, currentTick time.Time) bool {
+	// Don't attempt to autostop failed workspaces.
+	if !job.CompletedAt.Valid || job.Error.String != "" {
+		return false
 	}
 
-	switch priorHistory.Transition {
-	case database.WorkspaceTransitionStart:
-		if priorHistory.Deadline.IsZero() {
-			return "", time.Time{}, xerrors.Errorf("latest workspace build has zero deadline")
-		}
-		// For stopping, do not truncate. This is inconsistent with autostart, but
-		// it ensures we will not stop too early.
-		return database.WorkspaceTransitionStop, priorHistory.Deadline, nil
-	case database.WorkspaceTransitionStop:
-		sched, err := schedule.Weekly(ws.AutostartSchedule.String)
-		if err != nil {
-			return "", time.Time{}, xerrors.Errorf("workspace has invalid autostart schedule: %w", err)
-		}
-		// Round down to the nearest minute, as this is the finest granularity cron supports.
-		// Truncate is probably not necessary here, but doing it anyway to be sure.
-		nextTransition = sched.Next(priorHistory.CreatedAt).Truncate(time.Minute)
-		return database.WorkspaceTransitionStart, nextTransition, nil
-	default:
-		return "", time.Time{}, xerrors.Errorf("last transition not valid for autostart or autostop")
-	}
+	// A workspace must be started in order for it to be auto-stopped.
+	return build.Transition == database.WorkspaceTransitionStart &&
+		!build.Deadline.IsZero() &&
+		// We do not want to stop a workspace prior to it breaching its deadline.
+		!currentTick.Before(build.Deadline)
+}
+
+// isEligibleForFailedStop returns true if the workspace is eligible to be stopped
+// due to a failed build.
```

Comment:
This should also check that it's a start job otherwise this will enter a loop of constantly stopping failed stop builds.